Re-instate the /days: option

noja45 opened the pull request:
Change the logic to "screen scrape" / analyse html in order to get prices directly from the Google Finance web pages.
This removes reliance on api's and re-enables the ability to get history (/days:n).
Added documents to give a bit more info on the html returned and tags to search for and a bit more info about the develompent.
A few "improvements";
  /symonly to cater better for differences between quicken and google market:symbol format
 ability to prefix symbols with "!" to ignore /div:n option for 1 symbol

avinash311 commented:
On error on one symbol, it does not save any other data.

And looks like currency rates would require a different way of scraping - for CADUSD:

Try 1, using Yahoo (non-Google) symbol 
Info: Current Price Treatment for symbol=CADUSD=X
WARN: Cannot find current price in getPrice() 'match(...).0' is null or not an object
C:\Users\avinash\github\QImportP\QImportP.wsf(342, 5) Microsoft JScript runtime error: 'match(...).0' is null or not an object

Using CADUSD, which has data on this URL: https://www.google.com/finance?q=usdcad
Info: Current Price Treatment for symbol=CADUSD
Warn: No current price for symbol CADUSD
QImportP: Done with script.

noja45 commented:
Hi avinash,
I still havent worked this GitHub thing out fully so I'll try to answer some of your comments;
1. I too am concerned about robustness of "screen scraping" - effectively it is the same as using a browser to get the html and then it will use regex match()/replace() to find/change specific tags (eg. class="pr" to find historic price table within html). I don't think it will get thrown off or require google login but it is possible that google may eventually decide to change the html tags so whan I search for "class=pr" it may change and not be found ... but whatever they change, it should be quite easy to change the regex to quickly cater for them ... better than removing an api which leaves nothing.
2. Yes, the 2 .doc's was more to explain to you (or any prospective developer) about my changes - 1 contains screen dumps and resulting html tags, the other about the development logic. 
As you say .doc is probably not a good idea and in reality they probably don't add anything so I will remove them.
3. Good testing, you found some problems;
a. Currency rates (eg. CADUSD) are not catered for [I didn't even think anyone would use currency rates in qImportP], though it is an easy to change to include them.
   [note. no history seems to be available for currency rates so only current currency rates can be found] 
b. A bigger problem is if the symbol isn't found, google will give 2 possibilities;
   i. Error - not found
   ii. A selection of possibilities.
   Which needs to be catered for in a better way in the script.
c. Seems like the error/info messages could be more informative to give a better clue as to any errors
   (to be honest, I just put a big block in try/catch to prevent the script aborting .. it needs to be better)
So, give me a few days, I'll try to change to cater for these issues.

Just for info. I'm the original Adrian who posted on aczoom about UK div100, use QHI to launch and (stupidly) InformExec!

Thanks ;-)

avinash311 commented:
Sounds good, I add more code comments once you have the updates submitted.
- handling missing/wrong ticker symbols (by skipping over those records, but saving other ticker data)
- change .doc to .txt, one or two such files, to document the explanations

In the next few weeks, I'm going to add one more method of creating current (only current, not historical) data by using Google Docs Spreadsheets which has a GOOGLEFINANCE function.
https://support.google.com/docs/answer/3093281?hl=en
(While it supports historical data for online spreadsheet, it does not allow export of spreadsheets using historical GOOGLEFINANCE calls.)

noja45 commented:
Hi again ... so I updated in my fork;
Removed 2 .docs - not really required
Better error information for "match" processing
Handles unknown symbols
Handles Currency rates

avinash311 commented on QImportP.wsf:
```diff
@@ -103,230 +100,551 @@ var QEXE_DEFAULT = Array(
+
+			// Never used because history url returns invalid symbol
+            // if webMarket = CURRENCY, there is no history 
+            //if (webMarket == "CURRENCY")
```
Does this if statement need to be uncommented, or at least enabled at DEBUG_LEVEL?

noja45 replied:
No, it is where I was catering for CURRENCY type having no history ... but Google doesn't return any values so the previous chack rejects and this code would never be executed for a CURRENCY type.
Comments now removed.

avinash311 commented on QImportP.wsf:
```diff
@@ -103,230 +100,551 @@ var QEXE_DEFAULT = Array(
+
+if (outFile)
+    {   
+	if (getPrices(symbols, numDays, outFilePath))
```
Function does not really use outFilePath, maybe better to pass outFile.

noja45 replied:
unused arguments are now removed

avinash311 commented on QImportP.wsf:
```diff
@@ -103,230 +100,551 @@ var QEXE_DEFAULT = Array(
+try  // open outut file and catch error, for example file already in use
+  {
+	// setup output file and write first line for quicken identification
+    var outFile = FSO.CreateTextFile(outFilePath);
```
Since outFile is used as a global var (and works in Javascript), make that clearer by moving the var outside this block.
Maybe
var outFile = null;
before the try.

noja45 replied:
now moved

avinash311 commented on QImportP.wsf:
```diff
@@ -103,230 +100,551 @@ var QEXE_DEFAULT = Array(
-
+function getPrices(symbols, numDays, outFilePath)
+    {
+    message("Start getPrices",ERROR_LEVEL);
```
Reduce to INFO_LEVEL or DEBUG_LEVEL, or maybe even delete this line,

noja45 replied:
Done

avinash311 commented on QImportP.wsf:
```diff
@@ -103,230 +100,551 @@ var QEXE_DEFAULT = Array(
+    WinHttpReq.SetTimeouts(30000, 30000, 30000, 30000)
+	
+	// setup domain parsor to process xml data
+	//var myDom = new ActiveXObject("MSXML2.DOMDocument.3.0");
```
Is this commented section useful?

noja45 replied:
no,  it is where I moved code but marked the original with comments - the idea being if the new location worked ok, I could delete the comments ..... provided I remembered to do so!

avinash311 commented on QImportP.wsf:
```diff
@@ -103,230 +100,551 @@ var QEXE_DEFAULT = Array(
+		if (symb.charAt(0) == "!")
+			{
+			dodiv = false;
+		    symb = symb.substr(1);
```
Would you be able to fix the indentation of all the new lines added?
The problems can been seen if you view the diff in the web browser in github, or any editor with tabstop=8 spaces (the default).
I suspect your editor is set to tabstop=4, so may look fine there.
Essentially, we need to eliminate and never use hard tabs in the source code, and expand tabs to 4 spaces in this change.

noja45 replied:
I have replaced all tab with 4xspace

avinash311 commented on QImportP.wsf:
```diff
@@ -103,230 +100,551 @@ var QEXE_DEFAULT = Array(
+
+		// current price for sym
+		var url = urlcurr.toUrl(sym);
+		WinHttpReq.Open("GET", url, false);
```
To help debug  this, add a DEBUG_LEVEL message that outputs the  url.

noja45 replied:
Added

noja45 replied:
Of course ... in order to debug I have "Microsoft Visual Studio (free version)" installed.
Calling " cscript //D //X  /s:%SYMBOLS%   %ARGUMENTS%   %*" invokes the VS debugger on start where I can set break-points, step through the code and inspect variables at various points. The message displays just give a better idea where to set the break-points

noja45 commented:
Regarding; cscript QImportP.wsf /s:C /v:3

    Start getPrices
    Info: Current Price Treatment for symbol=C
    C:\Users\avinash\github\QImportP\QImportP.wsf(449, 5) Microsoft JScript runtime error: 'match(...).0' is    null or not an object

I looked at this and it is because Market Indexs for America have different html tags when closed and the market is "After Hours" rather than "Closed".
While I was looking at it, I found the following html which seems to be common on all current prices regardless of  whether open,closed, mutf, stock or currency;
---->
<div id=sharebox-data     itemscope=itemscope     itemtype=http://schema.org/Intangible/FinancialQuote>
<meta itemprop=name        content=Citigroup Inc />
<meta itemprop=url        content=https://www.google.com/finance?cid=662713 />
<meta itemprop=imageUrl        content=https://www.google.com/finance/chart?cht=g&q=NYSE:C&tkr=1&p=1d&enddatetime=2017-08-23T16:01:42Z />
<meta itemprop=tickerSymbol        content=C />
<meta itemprop=exchange        content=NYSE />
<meta itemprop=exchangeTimezone        content=America/New_York />
<meta itemprop=price        content=67.23 />
<meta itemprop=priceChange        content=+0.38 />
<meta itemprop=priceChangePercent        content=0.57 />
<meta itemprop=isAfterHours        content=true>
<meta itemprop=afterHoursPrice        content=67.14>
<meta itemprop=afterHoursPriceChange        content=-0.09>
<meta itemprop=afterHoursPriceChangePercent        content=-0.13>
<meta itemprop=afterHoursQuoteTime        content=2017-08-23T23:53:22Z>
<meta itemprop=quoteTime        content=2017-08-23T16:01:42Z />
<meta itemprop=dataSource        content=NYSE real-time data />
<meta itemprop=dataSourceDisclaimerUrl        content=//www.google.com/help/stock_disclaimer.html#realtime />
<meta itemprop=priceCurrency        content=USD />
</div>
---------
Don't know why I didn't see this before as it makes life simpler so I've changed the current price/date logic;
1. use the <div id=sharebox-data> ..... </div>
2. find the itemprop="price" and use the content value as the price
3. find the itemprop="quoteTime" and the content value for the date
... if the <div sharebox-date> doesn't exist then either the user specified symbol is notfound/ambiguous and needs to change ... or google did an update and changed the id or something other.

Of course, a bit of testing for the new current price needs to be done .... I'll do a bit at my end too but have to wait for markets to open/close etc:- to try several possibilities

avinash311 commented:
Thanks, it worked for me now when I tested it, I did not realize it is still broken for after hours!

Did you also look at the &output=csv  argument that can be added to historical URL to get a spreadsheet? That seems to work now, and while some comments suggest that "Google Finance API" is going away, at least the web page seems to provide that access.
If that makes things easier, it could also be used for current prices, by doing a historical lookup for a day (or a week and picking latest price).

noja45 commented:
Sorry .... It isn't broken, it should work for "After Hours" too ...  
I did quite a bit of testing today and it all works for me, it was just a note to say a did a big change to the logic (which I am happier with as its a lot cleaner!) but I only spent a few hours testing. It seems to work ok but may need more tests (though I'm not expecting any fails).
